<a href="https://colab.research.google.com/github/sergiocostaifes/PPCOMP_DM/blob/main/notebooks/06_labeling_states.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# NB06 — Derivação de Estados e Formulação Supervisionada

## 1. Contexto

A partir da base temporal enriquecida produzida no NB05 e dos episódios críticos detectados
no NB04, este notebook formaliza a estrutura supervisionada do problema em nível de
janelas temporais de 5 minutos.

Nesta etapa, a série deixa de ser apenas uma sequência de atributos e episódios e passa a
ser organizada em:
- estados operacionais por janela;
- relações temporais entre janelas consecutivas;
- uma base supervisionada voltada à antecipação de entrada em estado crítico.

Como a base de features do NB05 já incorpora trim inicial por lags, este notebook precisa
considerar explicitamente que nem todos os episódios detectados na série scored completa
estão integralmente disponíveis na base final de atributos.

## 2. Objetivo

O objetivo deste notebook é:
- rotular cada janela temporal com um estado operacional;
- preservar coerência com os episódios críticos oficiais do pipeline;
- adaptar a rotulagem à faixa efetivamente disponível na base de features;
- derivar relações temporais entre janelas consecutivas;
- construir uma base supervisionada binária para antecipação de entrada em `DURING`
  dentro de um horizonte futuro fixo.

## 3. Papel no Pipeline Atualizado

Este notebook representa a passagem entre:
- a segmentação estatística de episódios;
- a engenharia de atributos temporais;
- e a formulação supervisionada utilizada pelos modelos posteriores.

Seu papel é transformar a série temporal enriquecida em dois artefatos principais:
- uma base rotulada por estados;
- uma base supervisionada de antecipação, adequada à modelagem binária posterior.

Assim, o NB06 consolida a transição entre a descrição temporal do comportamento da série
e a tarefa preditiva efetivamente avaliada no restante do pipeline.

## 4. Definição dos Estados

Cada janela é rotulada em uma das seguintes categorias:

- `DURING`: janela pertencente a um episódio crítico efetivamente disponível na base;
- `BEFORE`: janela situada até `K_BEFORE` passos antes do início de um episódio;
- `AFTER`: janela situada até `K_AFTER` passos após o fim de um episódio;
- `NORMAL`: janela fora das categorias anteriores.

A precedência adotada para evitar conflitos é:

1. `DURING`
2. `BEFORE` / `AFTER`
3. `NORMAL`

## 5. Ajuste entre Episódios Detectados e Base de Features

Os episódios detectados no NB04 referem-se à série scored completa. Já a base do NB05
pode não conter todas as janelas originais, devido ao trim inicial necessário para criação
dos lags.

Por isso, esta etapa trabalha com a noção de **episódios efetivos**:
- episódios integralmente contidos na faixa de buckets da base de features; ou
- episódios recortados pela interseção com essa faixa, quando necessário.

Esse procedimento evita inconsistências entre:
- o total de janelas críticas disponível na base de atributos;
- e a marcação `DURING` utilizada na formulação supervisionada.

## 6. Formulação Supervisionada

A formulação supervisionada construída nesta etapa é binária.

Para cada janela elegível no estado atual `NORMAL` ou `BEFORE`, define-se um horizonte
futuro fixo de `H = 12` janelas (60 minutos).

O alvo supervisionado é definido como:

- `1`: haverá pelo menos uma janela em estado `DURING` dentro do horizonte futuro;
- `0`: todo o horizonte futuro permanecerá em estado `NORMAL`.

Assim, a tarefa supervisionada não se restringe a transições imediatas do tipo
`BEFORE → DURING`, mas modela a antecipação de entrada em criticidade dentro de um
horizonte temporal futuro.

## 7. Relações Temporais Derivadas

Além do alvo binário, este notebook também deriva:
- `next_state`: estado da próxima janela;
- `transition`: transição nominal de um passo;
- flags auxiliares de validade do horizonte futuro;
- informações interpretativas sobre a trajetória futura de estados.

Esses campos são úteis para rastreabilidade, auditoria e interpretação dos resultados.

## 8. Saídas Esperadas

Ao final deste notebook, espera-se obter:
- base rotulada por estado (`window_5min_labeled.parquet`);
- base supervisionada binária de antecipação (`transition_dataset.parquet`);
- alias descritivo para a base supervisionada;
- resumo estruturado com distribuições de estados, elegibilidade e alvo.

## 9. Relação com as Próximas Etapas

Este notebook alimenta diretamente:
- NB08 — Modelagem de transições / antecipação;
- NB09 — Consolidação experimental e geração de figuras finais.

A qualidade da formulação supervisionada construída aqui impacta diretamente a validade
das etapas de modelagem e avaliação posteriores.

## 10. Observações Metodológicas

Esta etapa preserva causalidade temporal:
- os atributos vêm do NB05 e já foram construídos de forma causal;
- o alvo supervisionado é definido apenas por observação do horizonte futuro;
- a elegibilidade exclui janelas em `DURING` e `AFTER`, evitando ambiguidade semântica.

Além disso, a adaptação para episódios efetivos melhora a consistência entre a base de
atributos disponível e a rotulagem utilizada no problema supervisionado.

Este notebook consolida a formulação preditiva oficial do pipeline.

In [1]:
# ============================================================
# 06_labeling_states.ipynb
# Derivação de estados e formulação supervisionada (5-min)
# Episódios definidos por taxa (fail_rate)
# Pipeline PPCOMP_DM (Google Cluster Trace)
# ============================================================

# -----------------------------
# 0) BOOTSTRAP (Colab + Repo)
# -----------------------------
from pathlib import Path
import os
import sys
import subprocess
import importlib
import random
import json

import numpy as np
import pandas as pd
from IPython.display import display

SEED = 42
random.seed(SEED)
np.random.seed(SEED)

if not Path("/content/drive/MyDrive").exists():
    from google.colab import drive
    drive.mount("/content/drive")
else:
    print("[Bootstrap] Google Drive já montado.")

REPO_DIR = Path("/content/drive/MyDrive/Mestrado/PPCOMP_DM")
GITHUB_REPO = "https://github.com/sergiocostaifes/PPCOMP_DM.git"

if not REPO_DIR.exists():
    REPO_DIR.parent.mkdir(parents=True, exist_ok=True)
    subprocess.run(["git", "clone", GITHUB_REPO, str(REPO_DIR)], check=True)
else:
    try:
        subprocess.run(["git", "-C", str(REPO_DIR), "pull"], check=True)
    except Exception as e:
        print("[Bootstrap] Aviso no git pull:", e)

os.chdir(str(REPO_DIR))

repo_str = str(REPO_DIR)
if repo_str not in sys.path:
    sys.path.insert(0, repo_str)

importlib.invalidate_caches()

from src.paths import FEATURES_PATH, REPORTS_PATH, ensure_dirs
ensure_dirs()

def log(msg: str) -> None:
    print(f"[06_labeling_states] {msg}")

# -----------------------------
# 1) Metadados do NB05 e NB04
# -----------------------------
NB05_SUMMARY_FILE = REPORTS_PATH / "05_feature_engineering_summary.json"
NB04_SUMMARY_FILE = REPORTS_PATH / "04_detect_episodes_summary.json"

nb05_summary = {}
nb04_summary = {}

if NB05_SUMMARY_FILE.exists():
    nb05_summary = json.loads(NB05_SUMMARY_FILE.read_text(encoding="utf-8"))
    log(f"Resumo do NB05 carregado: {NB05_SUMMARY_FILE}")
else:
    log("Resumo do NB05 não encontrado; seguindo sem metadados adicionais.")

if NB04_SUMMARY_FILE.exists():
    nb04_summary = json.loads(NB04_SUMMARY_FILE.read_text(encoding="utf-8"))
    log(f"Resumo do NB04 carregado: {NB04_SUMMARY_FILE}")
else:
    log("Resumo do NB04 não encontrado; seguindo sem metadados adicionais.")

SCENARIO_LABEL = nb05_summary.get("scenario_label", nb04_summary.get("scenario_label", "unknown"))
REMOVE_HOUR_ZERO = nb05_summary.get("remove_hour_zero", nb04_summary.get("remove_hour_zero", None))
EXPECTED_FEATURE_ROWS = nb05_summary.get("rows_output", None)
EXPECTED_CRITICAL_WINDOWS_IN_FEATURES = nb05_summary.get("critical_windows_output", None)
EXPECTED_EPISODES_INPUT = nb05_summary.get("episodes_detected_input", None)

# -----------------------------
# 2) INPUTS
# -----------------------------
FEATURES_FILE = FEATURES_PATH / "window_5min_features.parquet"
EPISODES_FILE = FEATURES_PATH / "episodes_detected.parquet"

assert FEATURES_FILE.exists(), f"Arquivo não encontrado: {FEATURES_FILE}"
assert EPISODES_FILE.exists(), f"Arquivo não encontrado: {EPISODES_FILE}"

df = pd.read_parquet(FEATURES_FILE)
episodes = pd.read_parquet(EPISODES_FILE)

required_df_cols = {"bucket_id", "fail_rate", "is_critical"}
required_ep_cols = {"start_bucket", "end_bucket"}

missing_df = required_df_cols - set(df.columns)
missing_ep = required_ep_cols - set(episodes.columns)

assert not missing_df, f"Colunas ausentes em features: {missing_df}"
assert not missing_ep, f"Colunas ausentes em episodes: {missing_ep}"

df = df.sort_values("bucket_id").reset_index(drop=True).copy()
episodes = episodes.sort_values(["start_bucket", "end_bucket"]).reset_index(drop=True).copy()

log(f"Features shape: {df.shape}")
log(f"Episodes shape: {episodes.shape}")

if EXPECTED_FEATURE_ROWS is not None:
    assert int(len(df)) == int(EXPECTED_FEATURE_ROWS), (
        f"Quantidade de linhas em features difere do NB05: "
        f"esperado={EXPECTED_FEATURE_ROWS}, observado={len(df)}"
    )

# -----------------------------
# 3) PARÂMETROS
# -----------------------------
K_BEFORE = 12  # 60 min
K_AFTER = 12   # 60 min

STATE_NORMAL = "NORMAL"
STATE_BEFORE = "BEFORE"
STATE_DURING = "DURING"
STATE_AFTER = "AFTER"

# -----------------------------
# 4) VALIDAÇÃO DO EIXO TEMPORAL
# -----------------------------
is_sorted = bool(df["bucket_id"].is_monotonic_increasing)
has_unique_bucket = bool(df["bucket_id"].is_unique)
bucket_diff = df["bucket_id"].diff().dropna()

gap_counts = bucket_diff.value_counts().sort_index().to_dict()
n_gaps_gt1 = int((bucket_diff > 1).sum())

log(f"Eixo temporal ordenado por bucket_id: {is_sorted}")
log(f"bucket_id sem duplicidade: {has_unique_bucket}")
log(f"Número de lacunas internas (diff > 1): {n_gaps_gt1}")
log(f"Distribuição de diferenças entre buckets: {gap_counts}")

assert is_sorted, "bucket_id não está ordenado crescentemente."
assert has_unique_bucket, "Foram encontrados bucket_id duplicados."
assert n_gaps_gt1 == 0, "A base de features deveria estar contínua após o trim do NB05."

bucket_min_available = int(df["bucket_id"].min())
bucket_max_available = int(df["bucket_id"].max())

# -----------------------------
# 5) AJUSTE DOS EPISÓDIOS À FAIXA DISPONÍVEL
# -----------------------------
# Episódios detectados no NB04 podem incluir buckets que já não estão presentes
# em features por causa do trim inicial dos lags. Aqui construímos episódios efetivos.
effective_rows = []

for _, ep in episodes.iterrows():
    s = int(ep["start_bucket"])
    e = int(ep["end_bucket"])

    s_eff = max(s, bucket_min_available)
    e_eff = min(e, bucket_max_available)

    if s_eff <= e_eff:
        row = ep.to_dict()
        row["original_start_bucket"] = s
        row["original_end_bucket"] = e
        row["start_bucket"] = s_eff
        row["end_bucket"] = e_eff
        row["was_clipped"] = bool((s_eff != s) or (e_eff != e))
        row["duration_windows_effective"] = int(e_eff - s_eff + 1)
        effective_rows.append(row)

episodes_effective = pd.DataFrame(effective_rows)

episodes_input_total = int(len(episodes))
episodes_effective_total = int(len(episodes_effective))
episodes_removed_outside_features = int(episodes_input_total - episodes_effective_total)

log(f"Episodes input total: {episodes_input_total}")
log(f"Episodes effective total: {episodes_effective_total}")
log(f"Episodes removidos fora da faixa de features: {episodes_removed_outside_features}")

if EXPECTED_EPISODES_INPUT is not None:
    assert int(episodes_input_total) == int(EXPECTED_EPISODES_INPUT), (
        f"Total de episódios de entrada difere do esperado: "
        f"esperado={EXPECTED_EPISODES_INPUT}, observado={episodes_input_total}"
    )

# -----------------------------
# 6) INICIALIZAÇÃO DOS RÓTULOS
# -----------------------------
df["state"] = STATE_NORMAL

during_mask = np.zeros(len(df), dtype=bool)
before_mask = np.zeros(len(df), dtype=bool)
after_mask = np.zeros(len(df), dtype=bool)

bucket_values = df["bucket_id"].to_numpy()

# -----------------------------
# 7) MARCAÇÃO DURING
# -----------------------------
for _, ep in episodes_effective.iterrows():
    s = int(ep["start_bucket"])
    e = int(ep["end_bucket"])
    in_range = (bucket_values >= s) & (bucket_values <= e)
    during_mask |= in_range

df.loc[during_mask, "state"] = STATE_DURING

during_total = int(during_mask.sum())
log(f"DURING total: {during_total}")

# Validação contra is_critical disponível em features
critical_windows_features = int(df["is_critical"].sum())
log(f"Critical windows em features (is_critical): {critical_windows_features}")

assert during_total == critical_windows_features, (
    f"Inconsistência entre DURING e is_critical em features: "
    f"during_total={during_total}, is_critical_sum={critical_windows_features}"
)

if EXPECTED_CRITICAL_WINDOWS_IN_FEATURES is not None:
    assert during_total == int(EXPECTED_CRITICAL_WINDOWS_IN_FEATURES), (
        f"Total de DURING difere do NB05: "
        f"esperado={EXPECTED_CRITICAL_WINDOWS_IN_FEATURES}, observado={during_total}"
    )

# -----------------------------
# 8) MARCAÇÃO BEFORE / AFTER
# -----------------------------
for _, ep in episodes_effective.iterrows():
    s = int(ep["start_bucket"])
    e = int(ep["end_bucket"])

    before_range = (bucket_values >= s - K_BEFORE) & (bucket_values < s)
    after_range = (bucket_values > e) & (bucket_values <= e + K_AFTER)

    before_mask |= (before_range & ~during_mask)
    after_mask |= (after_range & ~during_mask)

df.loc[(df["state"] == STATE_NORMAL) & before_mask, "state"] = STATE_BEFORE
df.loc[(df["state"] == STATE_NORMAL) & after_mask, "state"] = STATE_AFTER

# -----------------------------
# 9) VALIDAÇÕES DA ROTULAGEM
# -----------------------------
dist_states = df["state"].value_counts().to_dict()

conf_before_during = int((before_mask & during_mask).sum())
conf_after_during = int((after_mask & during_mask).sum())

episodes_without_during = 0
for _, ep in episodes_effective.iterrows():
    s = int(ep["start_bucket"])
    e = int(ep["end_bucket"])
    hits = int(((df["bucket_id"] >= s) &
                (df["bucket_id"] <= e) &
                (df["state"] == STATE_DURING)).sum())
    if hits == 0:
        episodes_without_during += 1

log(f"Distribuição de estados: {dist_states}")
log(f"Conflitos BEFORE∩DURING: {conf_before_during}")
log(f"Conflitos AFTER∩DURING: {conf_after_during}")
log(f"Episódios efetivos sem DURING: {episodes_without_during}")

assert conf_before_during == 0, "Há conflito entre BEFORE e DURING."
assert conf_after_during == 0, "Há conflito entre AFTER e DURING."
assert episodes_without_during == 0, "Há episódios efetivos sem qualquer janela DURING."

expected_during_effective = int(episodes_effective["duration_windows_effective"].sum()) if len(episodes_effective) > 0 else 0
actual_during = int((df["state"] == STATE_DURING).sum())

log(f"DURING esperado (episódios efetivos): {expected_during_effective}")
log(f"DURING observado: {actual_during}")

assert actual_during == expected_during_effective, (
    f"Inconsistência no total de janelas DURING: "
    f"esperado={expected_during_effective}, observado={actual_during}"
)

# -----------------------------
# 10) DERIVAÇÃO DE next_state E transition (1 passo)
# -----------------------------
df["next_bucket_id"] = df["bucket_id"].shift(-1)
df["next_state"] = df["state"].shift(-1)

df["is_last_window"] = df["next_bucket_id"].isna()
df["has_valid_next"] = (
    (~df["is_last_window"]) &
    ((df["next_bucket_id"] - df["bucket_id"]) == 1)
)

df["transition"] = np.where(
    df["has_valid_next"],
    df["state"] + "->" + df["next_state"],
    pd.NA
)

# -----------------------------
# 11) DATASET SUPERVISIONADO DE ANTECIPAÇÃO
# Problema binário:
# 1 = entra em DURING dentro do horizonte H
# 0 = permanece NORMAL durante todo o horizonte H
# -----------------------------
HORIZON_WINDOWS = K_BEFORE  # 12 janelas = 60 min

state_by_bucket = dict(zip(df["bucket_id"], df["state"]))

future_valid_flags = []
future_has_during_flags = []
future_all_normal_flags = []
future_path_states = []
future_first_during_offset = []

for b in df["bucket_id"].tolist():
    future_states = []
    valid_horizon = True

    for h in range(1, HORIZON_WINDOWS + 1):
        next_b = b + h
        s = state_by_bucket.get(next_b, None)

        if s is None:
            valid_horizon = False
            break

        future_states.append(s)

    if not valid_horizon:
        future_valid_flags.append(False)
        future_has_during_flags.append(False)
        future_all_normal_flags.append(False)
        future_path_states.append(pd.NA)
        future_first_during_offset.append(pd.NA)
        continue

    has_during = any(s == STATE_DURING for s in future_states)
    all_normal = all(s == STATE_NORMAL for s in future_states)

    first_during = pd.NA
    if has_during:
        for idx, s in enumerate(future_states, start=1):
            if s == STATE_DURING:
                first_during = idx
                break

    future_valid_flags.append(True)
    future_has_during_flags.append(has_during)
    future_all_normal_flags.append(all_normal)
    future_path_states.append(" | ".join(future_states))
    future_first_during_offset.append(first_during)

df["future_horizon_valid"] = future_valid_flags
df["future_has_during"] = future_has_during_flags
df["future_all_normal"] = future_all_normal_flags
df["future_path_states"] = future_path_states
df["future_first_during_offset"] = future_first_during_offset

# Elegibilidade:
# - não usar DURING nem AFTER como estado atual
# - exigir horizonte futuro completo e contíguo
# - manter apenas casos claramente positivos ou claramente negativos
eligible_mask = (
    df["state"].isin([STATE_NORMAL, STATE_BEFORE]) &
    df["future_horizon_valid"] &
    (df["future_has_during"] | df["future_all_normal"])
)

transition_df = df.loc[eligible_mask].copy()

transition_df["transition_target"] = np.where(
    transition_df["future_has_during"], 1, 0
).astype("int64")

transition_df["transition_label"] = np.where(
    transition_df["transition_target"] == 1,
    f"TO_{STATE_DURING}_WITHIN_{HORIZON_WINDOWS}_WINDOWS",
    f"STAYS_{STATE_NORMAL}_{HORIZON_WINDOWS}_WINDOWS"
)

transition_df["prediction_horizon_minutes"] = HORIZON_WINDOWS * 5

# Sanidade
assert transition_df["transition_target"].isin([0, 1]).all(), \
    "Foram encontrados rótulos inválidos em transition_target."

dist_transition_target = (
    transition_df["transition_target"]
    .value_counts()
    .sort_index()
    .to_dict()
)

dist_transition_label = (
    transition_df["transition_label"]
    .value_counts()
    .to_dict()
)

dist_current_state_in_transition = (
    transition_df["state"]
    .value_counts()
    .to_dict()
)

eligible_rows = int(eligible_mask.sum())
future_valid_rows = int(df["future_horizon_valid"].sum())
future_has_during_rows = int(df["future_has_during"].sum())
future_all_normal_rows = int(df["future_all_normal"].sum())

log(f"Rows com horizonte válido: {future_valid_rows}")
log(f"Rows com future_has_during: {future_has_during_rows}")
log(f"Rows com future_all_normal: {future_all_normal_rows}")
log(f"Transition dataset shape: {transition_df.shape}")
log(f"Distribuição de transition_target: {dist_transition_target}")
log(f"Distribuição de transition_label: {dist_transition_label}")
log(f"Distribuição do estado atual no transition dataset: {dist_current_state_in_transition}")

# -----------------------------
# 12) PERSISTÊNCIA
# -----------------------------
OUT_LABELED_FILE = FEATURES_PATH / "window_5min_labeled.parquet"
OUT_LABELED_SCENARIO_FILE = FEATURES_PATH / f"window_5min_labeled_{SCENARIO_LABEL}.parquet"

OUT_TRANSITION_FILE = FEATURES_PATH / "transition_dataset.parquet"
OUT_TRANSITION_SCENARIO_FILE = FEATURES_PATH / f"transition_dataset_{SCENARIO_LABEL}.parquet"

# Alias descritivo, sem quebrar o pipeline atual
OUT_ANTICIPATION_FILE = FEATURES_PATH / "anticipation_supervised_dataset.parquet"
OUT_ANTICIPATION_SCENARIO_FILE = FEATURES_PATH / f"anticipation_supervised_dataset_{SCENARIO_LABEL}.parquet"

df.to_parquet(OUT_LABELED_FILE, compression="snappy", index=False)
df.to_parquet(OUT_LABELED_SCENARIO_FILE, compression="snappy", index=False)

transition_df.to_parquet(OUT_TRANSITION_FILE, compression="snappy", index=False)
transition_df.to_parquet(OUT_TRANSITION_SCENARIO_FILE, compression="snappy", index=False)

transition_df.to_parquet(OUT_ANTICIPATION_FILE, compression="snappy", index=False)
transition_df.to_parquet(OUT_ANTICIPATION_SCENARIO_FILE, compression="snappy", index=False)

log(f"Salvo labeled (canônico): {OUT_LABELED_FILE}")
log(f"Salvo labeled (cenário): {OUT_LABELED_SCENARIO_FILE}")
log(f"Salvo transition dataset (canônico): {OUT_TRANSITION_FILE}")
log(f"Salvo transition dataset (cenário): {OUT_TRANSITION_SCENARIO_FILE}")
log(f"Salvo anticipation dataset (canônico): {OUT_ANTICIPATION_FILE}")
log(f"Salvo anticipation dataset (cenário): {OUT_ANTICIPATION_SCENARIO_FILE}")

# -----------------------------
# 13) SUMMARY JSON
# -----------------------------
summary = {
    "input_features_file": str(FEATURES_FILE),
    "input_episodes_file": str(EPISODES_FILE),

    "output_labeled_file_canonical": str(OUT_LABELED_FILE),
    "output_labeled_file_scenario": str(OUT_LABELED_SCENARIO_FILE),

    "output_transition_file_canonical": str(OUT_TRANSITION_FILE),
    "output_transition_file_scenario": str(OUT_TRANSITION_SCENARIO_FILE),

    "output_anticipation_file_canonical": str(OUT_ANTICIPATION_FILE),
    "output_anticipation_file_scenario": str(OUT_ANTICIPATION_SCENARIO_FILE),

    "scenario_label": SCENARIO_LABEL,
    "remove_hour_zero": REMOVE_HOUR_ZERO,

    "rows_labeled": int(len(df)),
    "cols_labeled": int(df.shape[1]),
    "rows_transition": int(len(transition_df)),
    "cols_transition": int(transition_df.shape[1]),

    "k_before": int(K_BEFORE),
    "k_after": int(K_AFTER),
    "horizon_windows": int(HORIZON_WINDOWS),
    "horizon_minutes": int(HORIZON_WINDOWS * 5),

    "is_bucket_id_sorted": bool(is_sorted),
    "is_bucket_id_unique": bool(has_unique_bucket),
    "n_internal_gaps_gt1": int(n_gaps_gt1),
    "bucket_diff_distribution": {str(k): int(v) for k, v in gap_counts.items()},

    "bucket_min_available": int(bucket_min_available),
    "bucket_max_available": int(bucket_max_available),

    "episodes_input_total": int(episodes_input_total),
    "episodes_effective_total": int(episodes_effective_total),
    "episodes_removed_outside_features": int(episodes_removed_outside_features),

    "states_distribution": {k: int(v) for k, v in dist_states.items()},
    "transition_target_distribution": {str(k): int(v) for k, v in dist_transition_target.items()},
    "transition_label_distribution": {k: int(v) for k, v in dist_transition_label.items()},
    "transition_current_state_distribution": {
        k: int(v) for k, v in dist_current_state_in_transition.items()
    },

    "future_horizon_valid_rows": int(future_valid_rows),
    "future_has_during_rows": int(future_has_during_rows),
    "future_all_normal_rows": int(future_all_normal_rows),
    "eligible_rows": int(eligible_rows),

    "episodes_without_during": int(episodes_without_during),
    "conflicts_before_during": int(conf_before_during),
    "conflicts_after_during": int(conf_after_during),

    "during_expected_from_effective_episodes": int(expected_during_effective),
    "during_observed": int(actual_during),
    "critical_windows_features_input": int(critical_windows_features),
}

summary_file = REPORTS_PATH / "06_labeling_states_summary.json"
summary_scenario_file = REPORTS_PATH / f"06_labeling_states_summary_{SCENARIO_LABEL}.json"

summary_file.write_text(
    json.dumps(summary, indent=2, ensure_ascii=False),
    encoding="utf-8"
)
summary_scenario_file.write_text(
    json.dumps(summary, indent=2, ensure_ascii=False),
    encoding="utf-8"
)

log(f"Resumo salvo (canônico): {summary_file}")
log(f"Resumo salvo (cenário): {summary_scenario_file}")
log("Notebook 06 finalizado com sucesso.")

# -----------------------------
# 14) AMOSTRAS DE CONFERÊNCIA
# -----------------------------
print("\n[Prévia da base rotulada]")
display(df.head())

print("\n[Prévia da base de transição / antecipação]")
display(transition_df.head())

print("\n[Distribuição final de estados]")
display(pd.Series(dist_states, name="count").to_frame())

print("\n[Distribuição final de transition_target]")
display(pd.Series(dist_transition_target, name="count").to_frame())

print("\n[Distribuição final de transition_label]")
display(pd.Series(dist_transition_label, name="count").to_frame())

print("\n[Distribuição do estado atual no transition dataset]")
display(pd.Series(dist_current_state_in_transition, name="count").to_frame())

Mounted at /content/drive
[06_labeling_states] Resumo do NB05 carregado: /content/drive/MyDrive/Mestrado/04-reports/05_feature_engineering_summary.json
[06_labeling_states] Resumo do NB04 carregado: /content/drive/MyDrive/Mestrado/04-reports/04_detect_episodes_summary.json
[06_labeling_states] Features shape: (8927, 28)
[06_labeling_states] Episodes shape: (285, 17)
[06_labeling_states] Eixo temporal ordenado por bucket_id: True
[06_labeling_states] bucket_id sem duplicidade: True
[06_labeling_states] Número de lacunas internas (diff > 1): 0
[06_labeling_states] Distribuição de diferenças entre buckets: {1.0: 8926}
[06_labeling_states] Episodes input total: 285
[06_labeling_states] Episodes effective total: 284
[06_labeling_states] Episodes removidos fora da faixa de features: 1
[06_labeling_states] DURING total: 389
[06_labeling_states] Critical windows em features (is_critical): 389
[06_labeling_states] Distribuição de estados: {'NORMAL': 5572, 'BEFORE': 1987, 'AFTER': 979, 'DURING':

,bucket_id,bucket_start_us,n_events,n_failed,n_machines,n_collections,mean_priority,mean_req_cpus,mean_req_mem,req_cpus_presence_rate,...,next_bucket_id,next_state,is_last_window,has_valid_next,transition,future_horizon_valid,future_has_during,future_all_normal,future_path_states,future_first_during_offset
0,3,900000000,32,2,32,18,135.656250,0.009495,0.003928,1.0,...,4.0,NORMAL,False,True,NORMAL->NORMAL,True,False,True,NORMAL | NORMAL | NORMAL | NORMAL | NORMAL | N...,<NA>
1,4,1200000000,29,6,28,20,233.068970,0.012932,0.003717,1.0,...,5.0,NORMAL,False,True,NORMAL->NORMAL,True,False,True,NORMAL | NORMAL | NORMAL | NORMAL | NORMAL | N...,<NA>
2,5,1500000000,25,5,25,19,253.360001,0.011083,0.012531,1.0,...,6.0,NORMAL,False,True,NORMAL->NORMAL,True,False,True,NORMAL | NORMAL | NORMAL | NORMAL | NORMAL | N...,<NA>
3,6,1800000000,25,4,25,21,241.559998,0.011335,0.007542,1.0,...,7.0,NORMAL,False,True,NORMAL->NORMAL,True,False,True,NORMAL | NORMAL | NORMAL | NORMAL | NORMAL | N...,<NA>
4,7,2100000000,40,12,37,25,249.399994,0.016829,0.006852,1.0,...,8.0,NORMAL,False,True,NORMAL->NORMAL,True,False,True,NORMAL | NORMAL | NORMAL | NORMAL | NORMAL | N...,<NA>



[Prévia da base de transição / antecipação]


,bucket_id,bucket_start_us,n_events,n_failed,n_machines,n_collections,mean_priority,mean_req_cpus,mean_req_mem,req_cpus_presence_rate,...,has_valid_next,transition,future_horizon_valid,future_has_during,future_all_normal,future_path_states,future_first_during_offset,transition_target,transition_label,prediction_horizon_minutes
0,3,900000000,32,2,32,18,135.656250,0.009495,0.003928,1.0,...,True,NORMAL->NORMAL,True,False,True,NORMAL | NORMAL | NORMAL | NORMAL | NORMAL | N...,<NA>,0,STAYS_NORMAL_12_WINDOWS,60
1,4,1200000000,29,6,28,20,233.068970,0.012932,0.003717,1.0,...,True,NORMAL->NORMAL,True,False,True,NORMAL | NORMAL | NORMAL | NORMAL | NORMAL | N...,<NA>,0,STAYS_NORMAL_12_WINDOWS,60
2,5,1500000000,25,5,25,19,253.360001,0.011083,0.012531,1.0,...,True,NORMAL->NORMAL,True,False,True,NORMAL | NORMAL | NORMAL | NORMAL | NORMAL | N...,<NA>,0,STAYS_NORMAL_12_WINDOWS,60
3,6,1800000000,25,4,25,21,241.559998,0.011335,0.007542,1.0,...,True,NORMAL->NORMAL,True,False,True,NORMAL | NORMAL | NORMAL | NORMAL | NORMAL | N...,<NA>,0,STAYS_NORMAL_12_WINDOWS,60
4,7,2100000000,40,12,37,25,249.399994,0.016829,0.006852,1.0,...,True,NORMAL->NORMAL,True,False,True,NORMAL | NORMAL | NORMAL | NORMAL | NORMAL | N...,<NA>,0,STAYS_NORMAL_12_WINDOWS,60



[Distribuição final de estados]


,count
NORMAL,5572
BEFORE,1987
AFTER,979
DURING,389



[Distribuição final de transition_target]


,count
0,4868
1,1987



[Distribuição final de transition_label]


,count
STAYS_NORMAL_12_WINDOWS,4868
TO_DURING_WITHIN_12_WINDOWS,1987



[Distribuição do estado atual no transition dataset]


,count
NORMAL,4868
BEFORE,1987


## 11. Conclusão da Etapa

A etapa de derivação de estados e formulação supervisionada foi concluída com sucesso,
produzindo dois artefatos centrais do pipeline:

- uma base temporal rotulada por estados operacionais (`window_5min_labeled.parquet`);
- uma base supervisionada binária de antecipação (`transition_dataset.parquet`), também
  salva sob alias descritivo (`anticipation_supervised_dataset.parquet`).

Nesta execução, a formulação passou a considerar explicitamente a faixa efetivamente
disponível na base de features, ajustando os episódios detectados no NB04 ao intervalo
presente em `window_5min_features.parquet`.

Com isso, a etapa manteve coerência com:
- a segmentação crítica oficial do NB04;
- o trim inicial realizado no NB05;
- a formulação supervisionada binária usada nos notebooks posteriores.

## 12. Achados Experimentais da Execução Atual

A execução atual consolidou de forma mais robusta a passagem entre a descrição temporal
da série e a formulação preditiva supervisionada.

### 12.1 Estrutura da base de entrada

A base de features utilizada nesta execução apresentou:
- `rows_labeled = 8927`
- `cols_labeled = 39`
- `bucket_min_available = 3`
- `bucket_max_available = 8929`

Além disso:
- `bucket_id` está ordenado;
- `bucket_id` é único;
- não há lacunas internas (`n_internal_gaps_gt1 = 0`);
- a distribuição de diferenças entre buckets foi integralmente `1.0`.

Esses resultados confirmam que a base proveniente do NB05 está temporalmente íntegra para
a formulação supervisionada.

### 12.2 Ajuste entre episódios detectados e base disponível

O conjunto original de episódios carregado do NB04 continha:
- `episodes_input_total = 285`

Após o ajuste para a faixa efetivamente disponível na base de features:
- `episodes_effective_total = 284`
- `episodes_removed_outside_features = 1`

Esse resultado era esperado, pois o trim inicial do NB05 removeu as primeiras janelas da
série, tornando um dos episódios originais totalmente indisponível na base final de
atributos.

A adoção do conceito de episódios efetivos resolveu corretamente essa assimetria entre:
- a série scored completa do NB04;
- e a base efetiva de features do NB05.

### 12.3 Coerência entre `DURING` e criticidade disponível em features

A nova rotulagem produziu:
- `DURING total = 389`

Esse valor coincide exatamente com:
- `critical_windows_features_input = 389`
- `during_expected_from_effective_episodes = 389`
- `during_observed = 389`

Também foi verificado que:
- `episodes_without_during = 0`
- `conflicts_before_during = 0`
- `conflicts_after_during = 0`

Esses resultados mostram que a marcação dos estados ficou internamente consistente e
plenamente alinhada à base de features disponível nesta execução.

### 12.4 Distribuição final de estados

A rotulagem da base completa produziu a seguinte distribuição:

- `NORMAL = 5572`
- `BEFORE = 1987`
- `AFTER = 979`
- `DURING = 389`

Essa distribuição evidencia que:
- a maior parte das janelas permanece em regime normal;
- há uma quantidade expressiva de janelas `BEFORE`, o que é importante para a tarefa de
  antecipação;
- as janelas `DURING` refletem exatamente a criticidade remanescente na base após o trim.

### 12.5 Validade do horizonte futuro

Na etapa de formulação supervisionada, foram observados os seguintes totais:

- `future_horizon_valid_rows = 8915`
- `future_has_during_rows = 2273`
- `future_all_normal_rows = 4919`
- `eligible_rows = 6855`

Isso mostra que:
- quase toda a base ainda permite horizonte futuro completo de 12 janelas;
- uma parcela significativa observa criticidade futura dentro do horizonte;
- outra parcela igualmente relevante mantém normalidade completa no horizonte;
- a elegibilidade final restringe corretamente os casos utilizados no problema binário.

### 12.6 Estrutura da base supervisionada

A base supervisionada final apresentou:
- `rows_transition = 6855`
- `cols_transition = 42`

A distribuição do alvo supervisionado ficou:

- `transition_target_distribution = {0: 4868, 1: 1987}`

Ou seja:
- `0`: permanece normal durante todo o horizonte de 12 janelas;
- `1`: entra em `DURING` dentro do horizonte de 12 janelas.

A distribuição textual dos rótulos ficou:

- `STAYS_NORMAL_12_WINDOWS = 4868`
- `TO_DURING_WITHIN_12_WINDOWS = 1987`

Esse resultado confirma que a formulação supervisionada está operacionalmente coerente e
sem ambiguidades na codificação do alvo.

### 12.7 Estado atual das amostras elegíveis

A distribuição do estado atual dentro da base supervisionada ficou:

- `NORMAL = 4868`
- `BEFORE = 1987`

Isso mostra que a elegibilidade implementada foi respeitada:
- apenas janelas em `NORMAL` ou `BEFORE` entram no problema supervisionado;
- janelas em `DURING` e `AFTER` ficam fora da base de antecipação.

Esse ponto é importante porque confirma que a tarefa supervisionada não é simplesmente
uma classificação do estado atual, mas uma antecipação de criticidade futura a partir de
contextos elegíveis.

## 13. Importância para o Pipeline Analítico

O NB06 consolida a formulação preditiva oficial do pipeline.

Seu papel é:
- transformar a série enriquecida em uma base rotulada por estados;
- adaptar a segmentação episódica à base efetivamente disponível;
- construir o problema supervisionado binário usado nos modelos posteriores.

Nesta execução, o notebook melhora significativamente a consistência metodológica do
pipeline ao reconhecer explicitamente que:
- os episódios do NB04 foram detectados sobre uma série mais ampla;
- a base de features do NB05 já sofreu trim inicial;
- a rotulagem supervisionada deve ser construída sobre os episódios efetivamente
  disponíveis, e não apenas sobre a lista bruta de episódios detectados.

Isso evita inconsistências silenciosas e fortalece a defesa científica da formulação adotada.

## 14. Artefatos Gerados

Nesta execução, foram gerados os seguintes artefatos:

- `window_5min_labeled.parquet`
- `window_5min_labeled_keep_hour0.parquet`
- `transition_dataset.parquet`
- `transition_dataset_keep_hour0.parquet`
- `anticipation_supervised_dataset.parquet`
- `anticipation_supervised_dataset_keep_hour0.parquet`
- `06_labeling_states_summary.json`
- `06_labeling_states_summary_keep_hour0.json`

Esses arquivos passam a constituir a referência oficial da etapa de rotulagem e formulação
supervisionada para o cenário `keep_hour0`.

## 15. Encaminhamento do Pipeline

Com a base rotulada e a base supervisionada de antecipação devidamente persistidas, o
pipeline segue para:

- NB08 — Modelagem e avaliação dos algoritmos;
- NB09 — Consolidação experimental e geração de figuras finais.

A partir deste ponto, a modelagem supervisionada passa a operar sobre uma formulação
mais precisa, mais rastreável e mais coerente com a base efetivamente disponível após a
engenharia de atributos.